include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_histogram.h

```diff
@@ -0,0 +1,681 @@
+    }
+};
+
+// Specialization for custom range binhash function which stores boundary data
+// into SLM for quick repeated usage
+template <typename _Range, typename _ExtraMemAccessor>
+struct __binhash_SLM_wrapper<oneapi::dpl::__internal::__custom_range_binhash<_Range>, _ExtraMemAccessor>
+{
+    using _BinHashType = typename oneapi::dpl::__internal::__custom_range_binhash<_Range>;
+
+    _BinHashType __bin_hash;
+    _ExtraMemAccessor __slm_mem;
+    __binhash_SLM_wrapper(_BinHashType __bin_hash_, _ExtraMemAccessor __slm_mem_, const sycl::nd_item<1>& __self_item)
+        : __bin_hash(__bin_hash_), __slm_mem(__slm_mem_)
+    {
+        //initialize __slm_memory
+        ::std::uint32_t __gSize = __self_item.get_local_range()[0];
+        ::std::uint32_t __self_lidx = __self_item.get_local_id(0);
+        ::std::uint8_t __factor = oneapi::dpl::__internal::__dpl_ceiling_div(__bin_hash.__boundaries.size(), __gSize);
+        ::std::uint8_t __k = 0;
+        for (; __k < __factor - 1; ++__k)
+        {
+            __slm_mem[__gSize * __k + __self_lidx] = __bin_hash.__boundaries[__gSize * __k + __self_lidx];
+        }
+        // residual
+        if (__gSize * __k + __self_lidx < __bin_hash.__boundaries.size())
+        {
+            __slm_mem[__gSize * __k + __self_lidx] = __bin_hash.__boundaries[__gSize * __k + __self_lidx];
+        }
+    }
+
+    template <typename _T>
+    ::std::uint32_t
+    get_bin(_T&& __value) const
+    {
+        return _BinHashType::get_bin_helper(__slm_mem.begin(), __slm_mem.end(), ::std::forward<_T>(__value));
+    }
+
+    template <typename _T>
+    bool
+    is_valid(_T&& __value) const
+    {
+        return _BinHashType::is_valid_helper(__slm_mem[0], __slm_mem[__slm_mem.size() - 1],
+                                             ::std::forward<_T>(__value));
+    }
+};
+
+template <typename _BinHash, typename _ExtraMemAccessor>
+auto
+__make_SLM_binhash(_BinHash __bin_hash, _ExtraMemAccessor __slm_mem, const sycl::nd_item<1>& __self_item)
+{
+    return __binhash_SLM_wrapper(__bin_hash, __slm_mem, __self_item);
+}
+
+template <typename _BinHash>
+struct __binhash_manager_base
+{
+    //will always be empty, but just to have some type
+    using extra_memory_type = typename ::std::uint8_t;
+    _BinHash __bin_hash;
+    __binhash_manager_base(_BinHash __bin_hash_) : __bin_hash(__bin_hash_) {}
+
+    ::std::size_t
+    get_required_SLM_elements() const
+    {
+        return 0;
+    }
+
+    void
+    require_access(sycl::handler& __cgh)
+    {
+    }
+
+    auto
+    get_device_copyable_binhash()
+    {
+        return __bin_hash;
+    }
+};
+
+// Augmentation for binhash function which stores a range of dynamic memory to determine bin mapping
+template <typename _BinHash, typename _BufferType>
+struct __binhash_manager_with_buffer : __binhash_manager_base<_BinHash>
+{
+    using BaseType = __binhash_manager_base<_BinHash>;
+    using extra_memory_type = typename _BinHash::range_value_type;
```

Comment:
Some typedefs in this file are in `snake_case` and others are in `PascalCase`. In this particular instance, we are using different case conventions for two lines that are right next to eachother.

Should we decide on a single style for these typedefs?

In either case, both of these typedefs probably should be prefixed with an underscore.

Reply:
Thanks, this is a good point.  I've improved the naming of some of these, and I am using snake case with the leading `_`.